btree/btree.c

```diff
@@ -2377,6 +2411,288 @@ void m0_btree_ut_node_add_del_rec(void)
 	M0_LEAVE();
 }
 
+/**
+ * This unit test exercises a few tree operations in both valid and invalid
+ * conditions.
+ */
+static void m0_btree_ut_basic_tree_operations(void)
+{
+	void                   *invalid_addr = (void *)0xbadbadbadbad;
+	struct m0_btree        *btree;
+	struct m0_btree_type    btree_type = {.tt_id = M0_BT_EMAP_EM_MAPPING };
+	struct m0_be_tx        *tx = NULL;
+	struct m0_btree_op      b_op;
+	void                   *temp_node;
+
+	/** Prepare transaction to capture tree operations. */
+	m0_be_tx_init(tx, 0, NULL, NULL, NULL, NULL, NULL, NULL);
+	m0_be_tx_prep(tx, NULL);
+
+	/**
+	 *  Run a valid scenario:
+	 *  1) Create a btree
+	 *  2) Close the btree
+	 *  3) Open the btree
+	 *  4) Close the btree
+	 *  5) Destroy the btree
+	 */
+
+	/** Create temp node space*/
+	temp_node = m0_alloc_aligned((1024 + sizeof(struct nd)), 10);
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op,
+			      m0_btree_create(temp_node, 1024, &btree_type,
+					      tx, &b_op));
+
+	m0_btree_close(b_op.bo_arbor);
+
+	m0_btree_open(temp_node, 1024, &btree);
+
+	m0_btree_close(btree);
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op, m0_btree_destroy(btree, &b_op));
+
+	m0_free_aligned(temp_node, (1024 + sizeof(struct nd)), 10);
+
+	/** Now run some invalid cases */
+
+	/** Open a non-existent btree */
+	m0_btree_open(invalid_addr, 1024, &btree);
+
+	/** Close a non-existent btree */
+	m0_btree_close(btree);
+
+	/** Destroy a non-existent btree */
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op, m0_btree_destroy(btree, &b_op));
+
+	/** Create a new btree */
+	temp_node = m0_alloc_aligned((1024 + sizeof(struct nd)), 10);
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op,
+			      m0_btree_create(temp_node, 1024, &btree_type,
+					      tx, &b_op));
+
+	/** Close it */
+	m0_btree_close(b_op.bo_arbor);
+
+	/** Try closing again */
+	m0_btree_close(b_op.bo_arbor);
+
+	/** Re-open it */
+	m0_btree_open(invalid_addr, 1024, &btree);
+
+	/** Open it again */
+	m0_btree_open(invalid_addr, 1024, &btree);
+
+	/** Destory it */
+	M0_BTREE_OP_SYNC_WITH(&b_op.bo_op, m0_btree_destroy(btree, &b_op));
+
+	/** Attempt to reopen the destroyed tree */
+	m0_btree_open(invalid_addr, 1024, &btree);
```

Comment:
Are we going to add any validation for return codes? UT should somehow verify the expected outputs of tree operations

Reply:
This is what I was thinking since I was not sure if we would be asserting or returning error codes. We need to look at the current implementation if the current btree code asserts or returns error codes.